File: apprefurbished/src/main/java/com/androidcollider/easyfin/common/repository/Repository.java
package com.androidcollider.easyfin.common.repository;

import com.androidcollider.easyfin.common.models.Account;
import com.androidcollider.easyfin.common.models.Debt;
import com.androidcollider.easyfin.common.models.Rates;
import com.androidcollider.easyfin.common.models.Transaction;
import com.androidcollider.easyfin.common.models.TransactionCategory;

import java.util.List;
import java.util.Map;

import io.reactivex.rxjava3.core.Flowable;

/**
 * @author Ihor Bilous
 */

public interface Repository {

    // Accounts

    Flowable<Account> addNewAccount(Account account);

    Flowable<List<Account>> getAllAccounts();

    Flowable<Account> updateAccount(Account account);

    Flowable<Boolean> deleteAccount(int id);

    Flowable<Boolean> transferBTWAccounts(int idAccount1,
                                          double accountAmount1,
                                          int idAccount2,
                                          double accountAmount2);

    Flowable<Boolean> setAllAccounts(List<Account> accountList);


    // Transactions

    Flowable<Transaction> addNewTransaction(Transaction transaction);

    Flowable<List<Transaction>> getAllTransactions();

    Flowable<Transaction> updateTransaction(Transaction transaction);

    Flowable<Boolean> updateTransactionDifferentAccounts(Transaction transaction,
                                                         double oldAccountAmount,
                                                         int oldAccountId);

    Flowable<Boolean> deleteTransaction(int idAccount,
                                        int idTransaction,
                                        double amount);

    Flowable<Boolean> setAllTransactions(List<Transaction> transactionList);


    // Debts

    Flowable<Debt> addNewDebt(Debt debt);

    Flowable<List<Debt>> getAllDebts();

    Flowable<Debt> updateDebt(Debt debt);

    Flowable<Boolean> updateDebtDifferentAccounts(Debt debt,
                                                  double oldAccountAmount,
                                                  int oldAccountId);

    Flowable<Boolean> deleteDebt(int idAccount,
                                 int idDebt,
                                 double amount,
                                 int type);

    Flowable<Boolean> payFullDebt(int idAccount,
                                  double accountAmount,
                                  int idDebt);

    Flowable<Boolean> payPartOfDebt(int idAccount,
                                    double accountAmount,
                                    int idDebt,
                                    double debtAmount);

    Flowable<Boolean> takeMoreDebt(int idAccount,
                                   double accountAmount,
                                   int idDebt,
                                   double debtAmount,
                                   double debtAllAmount);

    Flowable<Boolean> setAllDebts(List<Debt> debtList);


    // Statistic

    Flowable<Map<String, double[]>> getTransactionsStatistic(int position);

    Flowable<Map<String, double[]>> getAccountsAmountSumGroupByTypeAndCurrency();


    // Rates

    Flowable<Boolean> updateRates(List<Rates> ratesList);

    Flowable<double[]> getRates();

    Flowable<Boolean> setRates(double[] rates);


    // Transaction Category

    // Income

    Flowable<TransactionCategory> addNewTransactionIncomeCategory(TransactionCategory transactionCategory);

    Flowable<List<TransactionCategory>> getAllTransactionIncomeCategories();

    Flowable<TransactionCategory> updateTransactionIncomeCategory(TransactionCategory transactionCategory);

    Flowable<Boolean> deleteTransactionIncomeCategory(int id);

    Flowable<Boolean> setAllTransactionIncomeCategories(List<TransactionCategory> transactionCategoryList);


    // Expense

    Flowable<TransactionCategory> addNewTransactionExpenseCategory(TransactionCategory transactionCategory);

    Flowable<List<TransactionCategory>> getAllTransactionExpenseCategories();

    Flowable<TransactionCategory> updateTransactionExpenseCategory(TransactionCategory transactionCategory);

    Flowable<Boolean> deleteTransactionExpenseCategory(int id);

    Flowable<Boolean> setAllTransactionExpenseCategories(List<TransactionCategory> transactionCategoryList);
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/common/repository/Repository.kt
package com.androidcollider.easyfin.common.repository

import com.androidcollider.easyfin.common.models.*
import io.reactivex.rxjava3.core.Single

/**
 * @author Ihor Bilous
 */
interface Repository {
    // Accounts
    fun addNewAccount(account: Account): Single<Account>
    val allAccounts: Single<List<Account>>?
    fun updateAccount(account: Account): Single<Account>
    fun deleteAccount(id: Int): Single<Boolean>
    fun transferBTWAccounts(
        idAccount1: Int,
        accountAmount1: Double,
        idAccount2: Int,
        accountAmount2: Double
    ): Single<Boolean>

    fun setAllAccounts(accountList: List<Account>): Single<Boolean>

    // Transactions
    fun addNewTransaction(transaction: Transaction): Single<Transaction>
    val allTransactions: Single<List<Transaction>>?
    fun updateTransaction(transaction: Transaction): Single<Transaction>
    fun updateTransactionDifferentAccounts(
        transaction: Transaction,
        oldAccountAmount: Double,
        oldAccountId: Int
    ): Single<Boolean>

    fun deleteTransaction(
        idAccount: Int,
        idTransaction: Int,
        amount: Double
    ): Single<Boolean>

    fun setAllTransactions(transactionList: List<Transaction>): Single<Boolean>

    // Debts
    fun addNewDebt(debt: Debt): Single<Debt>
    val allDebts: Single<List<Debt>>?
    fun updateDebt(debt: Debt): Single<Debt>
    fun updateDebtDifferentAccounts(
        debt: Debt,
        oldAccountAmount: Double,
        oldAccountId: Int
    ): Single<Boolean>

    fun deleteDebt(
        idAccount: Int,
        idDebt: Int,
        amount: Double,
        type: Int
    ): Single<Boolean>

    fun payFullDebt(
        idAccount: Int,
        accountAmount: Double,
        idDebt: Int
    ): Single<Boolean>

    fun payPartOfDebt(
        idAccount: Int,
        accountAmount: Double,
        idDebt: Int,
        debtAmount: Double
    ): Single<Boolean>

    fun takeMoreDebt(
        idAccount: Int,
        accountAmount: Double,
        idDebt: Int,
        debtAmount: Double,
        debtAllAmount: Double
    ): Single<Boolean>

    fun setAllDebts(debtList: List<Debt>): Single<Boolean>

    // Statistic
    fun getTransactionsStatistic(position: Int): Single<Map<String, DoubleArray>>?
    val accountsAmountSumGroupByTypeAndCurrency: Single<Map<String, DoubleArray>>?

    // Rates
    fun updateRates(ratesList: List<Rates>): Single<Boolean>
    val rates: Single<DoubleArray>?
    fun setRates(rates: DoubleArray): Single<Boolean>

    // Transaction Category
    // Income
    fun addNewTransactionIncomeCategory(transactionCategory: TransactionCategory):
            Single<TransactionCategory>

    val allTransactionIncomeCategories: Single<List<TransactionCategory>>?
    fun updateTransactionIncomeCategory(transactionCategory: TransactionCategory):
            Single<TransactionCategory>

    fun deleteTransactionIncomeCategory(id: Int): Single<Boolean>
    fun setAllTransactionIncomeCategories(transactionCategoryList: List<TransactionCategory>):
            Single<Boolean>

    // Expense
    fun addNewTransactionExpenseCategory(transactionCategory: TransactionCategory):
            Single<TransactionCategory>

    val allTransactionExpenseCategories: Single<List<TransactionCategory>>?
    fun updateTransactionExpenseCategory(transactionCategory: TransactionCategory):
            Single<TransactionCategory>

    fun deleteTransactionExpenseCategory(id: Int): Single<Boolean>
    fun setAllTransactionExpenseCategories(transactionCategoryList: List<TransactionCategory>):
            Single<Boolean>
}